extensions/2.0/Khronos/KHR_materials_volume/README.md

```diff
@@ -0,0 +1,250 @@
+
+<figure style="text-align:center">
+<img src="./figures/transmission-translucency.png"/>
+<figcaption><em>Comparison of combining subsurface scattering with either transmission or translucency. Left: Rough transmission and subsurface scattering. Middle: Translucency and subsurface scattering. Right: Translucency without subsurface scattering using a thin-walled material. Colors are adjusted manually so that they look similar in the three configurations. This adjustment is needed in order to account for differences in distances and to minimize the impact of energy loss from the rough microfacet BTDF.</em></figcaption>
+</figure>
+
+## Implementation
+
+*This section is non-normative.*
+
+The microfacet BTDF f<sub>transmission</sub> defined in `KHR_materials_transmission` now takes refraction into account. That means that we to use Snell's law to compute the modified half vector:
+
+H = -normalize(*ior<sub>i</sub>* * *V* + *ior<sub>o</sub>* * *L*)
+
+*ior<sub>i</sub>* and *ior<sub>o</sub>* denote the index of refraction of the incident and transmitted side of the surface, respectively. *V* is the vector pointing to the camera, *L* points to the light. In a path tracer that starts rays at the camera, *V* corresponds to the incident side of the surface, which is the side of the medium with *ior<sub>i</sub>*.
+
+Incident and transmitted index of refraction have to be correctly set by the renderer, depending on whether light enters or leaves the object. An algorithm for tracking the IOR through overlapping objects is described by in [Schmidt and Budge (2002)](#SchmidtBudge2002).
+
+The microfacet BTDF also makes use of the modified half vector and the indices of refraction.
+
+f<sub>transmission</sub> = *T* * *baseColor* * (1 - *F<sub>R</sub>*) * (abs(dot(*L*, *H*)) * abs(dot(*V*, *H*))) / (abs(dot(*L*, *N*)) * abs(dot(*V*, *N*))) * (*ior<sub>o</sub>*<sup>2</sup> * *G<sub>R</sub>* * *D<sub>R</sub>*) / (*ior<sub>i</sub>* * dot(*V*, *H*) + *ior<sub>o</sub>* * dot(*L*, *H*))
+
+where *T* is the transmission percentage defined by `KHR_materials_transmission`. The F<sub>R</sub>, D<sub>R</sub>, and G<sub>R</sub> terms are the same as in the specular reflection except using the modified half vector *H* calculated from the refraction direction. See [Walter et al. (2007)](#Walter2007) for details.
```

Comment:
The `G` term does not depend on the half vector, so can we just use `G` rather than G<sub>R</sub>? Or should G<sub>R</sub> use a modified L?

Reply:
The description is a bit fuzzy and I am unsure how many details we should have here... G depends on the half vector to check if V and L are on the same side of the microsurface regarding H. See [eq. 34 in Walter 2007](https://www.cs.cornell.edu/~srm/publications/EGSR07-btdf.pdf). Maybe it's good enough to link the paper? Or would you prefer a more detailed explanation?

Reply:
So this is the positive characteristic function, where G returns 0 if V and H are on opposite sides of the surface? `clamp(sign(VdotH / NdotV), 0.0, 1.0);`
Appendix B does not mention this, so I think it would be helpful to include here if it is now important for transmission.

Reply:
Yes. If I am not mistaken, it is important for reflection and transmission, and it is part of the G term in either case. I am wondering why it's not part of Appendix B? Could be that Appendix B is just not precise in this regard, or it is assumed to be handled "outside" of G. [libbsdf for example](https://github.com/NVIDIA/MDL-SDK/blob/master/src/mdl/jit/libbsdf/libbsdf.cpp#L678) has an early out if VdotH or LdotH are on the "wrong" sides of the surface (were "wrong" depends on reflection/transmission and H).

In the end, it is just important to use the modified H when dealing with refraction, and the "normal" H when dealing with reflection.

Reply:
Perhaps it is worth a mention in #1717?  It doesn't appear that we are applying this term in the [sampleviewer](https://github.com/KhronosGroup/glTF-Sample-Viewer/blob/88b18169541ad039efcf370ca168ea69b6b048e3/src/shaders/brdf.glsl#L35) either.

Reply:
The step function is now part of #1717.

Reply:
@abwood is this thread resolved?

Reply:
Yes, this is now covered in Appendix B.